lib/fetch/index.js

```diff
@@ -727,34 +740,87 @@ function finalizeResponse (fetchParams, response) {
+      fetchParams.processResponseEndOfBody(response)
+    }
+  }
+
+  // 3. If fetchParams’s process response is non-null, then queue a fetch task to run fetchParams’s process response given response,
+  //    with fetchParams’s task destination.
   if (fetchParams.processResponse != null) {
     fetchParams.processResponse(response)
   }
 
-  // 2. If fetchParams’s process response consume is non-null, then:.
-  //    TODO
-  //    1. Let processBody given nullOrBytes be this step: run fetchParams’s
-  //    process response consume given response and nullOrBytes.on.
-  //    TODO
-  //    2. Let processBodyError be this step: run fetchParams’s process
-  //    response consume given response and failure.on.
-  //    TODO
-  //    3. If response’s body is null, then queue a fetch task to run
-  //    processBody given null, with fetchParams’s task destination.on.
-  //    TODO
-  //    4. Otherwise, fully read response’s body given processBody,
-  //    processBodyError, and fetchParams’s task destination.on.
-  //    TODO
-
-  // TODO (spec): The spec doesn't specify this but we need to
-  // terminate fetch if we have an error response.
-  if (response.type === 'error') {
-    context.terminate({ reason: response.error })
+  // 4. If response’s body is null, then run processResponseEndOfBody.
+  if (response.body == null) {
+    processResponseEndOfBody()
+  }
+
+  // 5. Otherwise:
+  else {
+    TransformStream ??= require('stream/web').TransformStream
+    // 1. Let transformStream be a new a TransformStream.
+
+    // 2. Let identityTransformAlgorithm be an algorithm which, given chunk, enqueues chunk in transformStream.
+    const identityTransformAlgorithm = (chunk, controller) => {
+      controller.enqueue(chunk)
+    }
+
+    const transformStream = new TransformStream({
+      // 3. Set up transformStream with transformAlgorithm set to identityTransformAlgorithm and
+      //    flushAlgorithm set to processResponseEndOfBody.
+      transform: identityTransformAlgorithm,
+      flush: processResponseEndOfBody
+    })
+
+    // 4. Set response’s body to the result of piping response’s body through transformStream.
+    // Khafra note: although the spec is unclear, only readable streams can be piped through
+    if (response.body != null && isReadable(response.body?.stream)) {
+      response.body = response.body.pipeThrough(transformStream)
+    }
+  }
+
+  // 6. If fetchParams’s process response consume body is non-null, then:
+  if (fetchParams.processResponseConsumeBody != null) {
+    // 1. Let processBody given nullOrBytes be this step: run fetchParams’s process response consume body given response and nullOrBytes.
+    const processBody = (nullOrBytes) => {
+      fetchParams.processResponseConsumeBody(response, nullOrBytes)
+    }
+
+    // 2. Let processBodyError be this step: run fetchParams’s process response consume body given response and failure.
+    const processBodyError = () => {
```

Comment:
```suggestion
    const processBodyError = (error) => {
```

Reply:
Maybe?

Reply:
I thought of doing that but the step above said "Let processBody given nullOrBytes" and this one doesn't name the params. Should I add a failure param anyways?

Reply:
>  process response consume body **given response and failure**

Reply:
Checking back, this seems to be a mis-wording with the spec and you are correct - "processBodyError must be an algorithm accepting an exception." 👍